core/server/api/users.js

```diff
@@ -150,35 +150,40 @@ users = {
 
             // Check permissions
             return canThis(options.context).edit.user(options.id).then(function () {
-                if (data.users[0].roles && data.users[0].roles[0]) {
-                    var role = data.users[0].roles[0],
-                        roleId = parseInt(role.id || role, 10);
-
-                    return dataProvider.User.findOne(
-                        {id: options.context.user, status: 'all'}, {include: ['roles']}
-                    ).then(function (contextUser) {
-                        var contextRoleId = contextUser.related('roles').toJSON()[0].id;
-
-                        if (roleId !== contextRoleId &&
-                                parseInt(options.id, 10) === parseInt(options.context.user, 10)) {
-                            return Promise.reject(new errors.NoPermissionError('You cannot change your own role.'));
-                        } else if (roleId !== contextRoleId) {
-                            return dataProvider.User.findOne({role: 'Owner'}).then(function (result) {
-                                if (parseInt(result.id, 10) !== parseInt(options.id, 10)) {
-                                    return canThis(options.context).assign.role(role).then(function () {
-                                        return editOperation();
-                                    });
-                                } else {
-                                    return Promise.reject(new errors.NoPermissionError('There has to be one owner.'));
+                // if roles aren't in the payload, proceed with the edit
+                if (!(data.users[0].roles && data.users[0].roles[0])) {
+                    return editOperation();
```

Comment:
I think so but if you're on to something to make it better I'm open to it :)

Right now we check the payload to make sure it's valid, check to make sure the context user has permission to edit the user (`canThis`), and then if roles are involved we do the additional checks for roles, and then we attempt the edit.

So if we don't need to worry about roles it seems pretty straightforward:  `canThis` user edit?  -> attempt edit.